### PR TITLE
Add BoatType Registry

### PR DESCRIPTION
**Pull request by cech12**
This PR adds a new registry for boat types. (like for example painting types)

It enables the possibility to add new boat types. For example, when a mod adds a new wood type, it can now use the new registry to add the new wood look also to the existing boat entity.
The abstract BoatType has two methods which the mod developer has to implement:
- `getBoatItem` where the boat item has to be returned
- `getOnBreakItems` where a list of items has to be returned, which are dropped, when the boat breaks.

The texture must be in `textures/entity/boat` and must have the name of the path of the ResourceLocation of the boat type.

The added test mod adds a green test boat.

Use Case: I want to create a mod with a new wood type and I want to add a new texture for boat entity.

### CODE REVIEW

**Comment by ichttt** [comment]
TBH I don't think that's a good idea, as it breaks binary compat, and creates many patches. You should probably just make BoatType an IExtensibleEnum, and patch the in getItemBoat to delegate to a method in BoatType

File: patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch [hunk 1]
```diff
@@ -30,3 +134,74 @@
     public static enum Status {
        IN_WATER,
        UNDER_WATER,
+@@ -786,18 +804,22 @@
+       IN_AIR;
+    }
+ 
+-   public static enum Type {
```
Comment (killjoy1221) [inline patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch]: This is what `EnumHelper` is for.

File: patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch [hunk 2]
```diff
@@ -1,6 +1,47 @@
 --- a/net/minecraft/entity/item/BoatEntity.java
 +++ b/net/minecraft/entity/item/BoatEntity.java
-@@ -459,7 +459,7 @@
+@@ -49,6 +49,7 @@
+    private static final DataParameter<Integer> field_184462_b = EntityDataManager.func_187226_a(BoatEntity.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Float> field_184464_c = EntityDataManager.func_187226_a(BoatEntity.class, DataSerializers.field_187193_c);
+    private static final DataParameter<Integer> field_184466_d = EntityDataManager.func_187226_a(BoatEntity.class, DataSerializers.field_187192_b);
++   private static final DataParameter<String> BOAT_TYPE_NAME = EntityDataManager.func_187226_a(BoatEntity.class, DataSerializers.field_187194_d);
```
Comment (MajorTuvok) [inline patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:7]: Are you sure vanilla doesn't have Problems with another Data Parameter...?

Comment (cech12) [reply patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:7]: I have checked a map generated without these changes. But with Forge. Maybe Vanilla acts different.

**Comment by cech12** [comment]
Thanks for your feedback. I will rework my changes. I like the idea with the IExtensibleEnum interface.

**Comment by cech12** [comment]
I added a new pull request #6263

**Pull request closed without merging**